data race in simulator/view_manager

Found while goroutines were using the same ViewManager

```
WARNING: DATA RACE
Write at 0x00c00050c0a0 by goroutine 94:
  github.com/vmware/govmomi/simulator.(*ViewManager).CreateContainerView()
      C:/Users/toomasp/IdeaProjects/vcenter_scanner.go/vendor/github.com/vmware/govmomi/simulator/view_manager.go:122 +0x737

Previous read at 0x00c00050c0a0 by goroutine 25:
  github.com/vmware/govmomi/simulator.(*ViewManager).Reference()
      <autogenerated>:1 +0x74
  github.com/vmware/govmomi/simulator.(*Registry).locker()
      C:/Users/toomasp/IdeaProjects/vcenter_scanner.go/vendor/github.com/vmware/govmomi/simulator/registry.go:505 +0xf2
```

Did, appears that it doesn't trigger reliably enough ( sorry for that, but it seems that I just lucked out when testing it )

```bash
toomasp@playtech-v1914  ~/tmp/govmomi  ➦ 9b0db1c  GORACE=history_size=5 go test -race -count 1 -v ./simulator -test.run ^TestViewManager_CreateContainerView$
=== RUN   TestViewManager_CreateContainerView
--- PASS: TestViewManager_CreateContainerView (1.05s)
PASS
ok  	github.com/vmware/govmomi/simulator	2.139s
 toomasp@playtech-v1914  ~/tmp/govmomi  ➦ 9b0db1c  GORACE=history_size=5 go test -race -count 1 -v ./simulator -test.run ^TestViewManager_CreateContainerView$
=== RUN   TestViewManager_CreateContainerView
==================
WARNING: DATA RACE
Read at 0x00c000032c20 by goroutine 76:
  github.com/vmware/govmomi/simulator.(*ViewManager).Reference()
      <autogenerated>:1 +0x6d
  github.com/vmware/govmomi/simulator.(*Registry).locker()
      /home/toomasp/tmp/govmomi/simulator/registry.go:505 +0xeb
  github.com/vmware/govmomi/simulator.(*Registry).WithLock()
      /home/toomasp/tmp/govmomi/simulator/registry.go:528 +0x88
  github.com/vmware/govmomi/simulator.(*Service).call()
      /home/toomasp/tmp/govmomi/simulator/simulator.go:208 +0x5c0
  github.com/vmware/govmomi/simulator.(*Service).ServeSDK()
      /home/toomasp/tmp/govmomi/simulator/simulator.go:439 +0x16b5
  github.com/vmware/govmomi/simulator.(*Service).ServeSDK-fm()
      /home/toomasp/tmp/govmomi/simulator/simulator.go:399 +0x5f
  net/http.HandlerFunc.ServeHTTP()
      /opt/go/src/net/http/server.go:1995 +0x51
  net/http.(*ServeMux).ServeHTTP()
      /opt/go/src/net/http/server.go:2375 +0x28a
  net/http.serverHandler.ServeHTTP()
      /opt/go/src/net/http/server.go:2774 +0xce
  net/http.(*conn).serve()
      /opt/go/src/net/http/server.go:1878 +0x811

Previous write at 0x00c000032c20 by goroutine 77:
  github.com/vmware/govmomi/simulator.(*ViewManager).CreateContainerView()
      /home/toomasp/tmp/govmomi/simulator/view_manager.go:122 +0x73b
  runtime.call64()
      /opt/go/src/runtime/asm_amd64.s:520 +0x3a
  reflect.Value.Call()
      /opt/go/src/reflect/value.go:308 +0xc0
  github.com/vmware/govmomi/simulator.(*Service).call.func1()
      /home/toomasp/tmp/govmomi/simulator/simulator.go:209 +0x7b
  github.com/vmware/govmomi/simulator.(*Registry).WithLock()
      /home/toomasp/tmp/govmomi/simulator/registry.go:532 +0x51
  github.com/vmware/govmomi/simulator.(*Service).call()
      /home/toomasp/tmp/govmomi/simulator/simulator.go:208 +0x5c0
  github.com/vmware/govmomi/simulator.(*Service).ServeSDK()
      /home/toomasp/tmp/govmomi/simulator/simulator.go:439 +0x16b5
  github.com/vmware/govmomi/simulator.(*Service).ServeSDK-fm()
      /home/toomasp/tmp/govmomi/simulator/simulator.go:399 +0x5f
  net/http.HandlerFunc.ServeHTTP()
      /opt/go/src/net/http/server.go:1995 +0x51
  net/http.(*ServeMux).ServeHTTP()
      /opt/go/src/net/http/server.go:2375 +0x28a
  net/http.serverHandler.ServeHTTP()
      /opt/go/src/net/http/server.go:2774 +0xce
  net/http.(*conn).serve()
      /opt/go/src/net/http/server.go:1878 +0x811

Goroutine 76 (running) created at:
  net/http.(*Server).Serve()
      /opt/go/src/net/http/server.go:2884 +0x4c4
  github.com/vmware/govmomi/simulator/internal.(*Server).goServe.func1()
      /home/toomasp/tmp/govmomi/simulator/internal/server.go:268 +0xac

Goroutine 77 (running) created at:
  net/http.(*Server).Serve()
      /opt/go/src/net/http/server.go:2884 +0x4c4
  github.com/vmware/govmomi/simulator/internal.(*Server).goServe.func1()
      /home/toomasp/tmp/govmomi/simulator/internal/server.go:268 +0xac
==================
==================
WARNING: DATA RACE
Read at 0x00c000776020 by goroutine 54:
  github.com/vmware/govmomi/simulator.RemoveReference()
      /home/toomasp/tmp/govmomi/simulator/registry.go:394 +0x90
  github.com/vmware/govmomi/simulator.destroyView()
      /home/toomasp/tmp/govmomi/simulator/view_manager.go:74 +0x8a
  github.com/vmware/govmomi/simulator.(*ContainerView).DestroyView()
      /home/toomasp/tmp/govmomi/simulator/view_manager.go:144 +0x105
  runtime.call64()
      /opt/go/src/runtime/asm_amd64.s:520 +0x3a
  reflect.Value.Call()
      /opt/go/src/reflect/value.go:308 +0xc0
  github.com/vmware/govmomi/simulator.(*Service).call.func1()
      /home/toomasp/tmp/govmomi/simulator/simulator.go:209 +0x7b
  github.com/vmware/govmomi/simulator.(*Registry).WithLock()
      /home/toomasp/tmp/govmomi/simulator/registry.go:532 +0x51
  github.com/vmware/govmomi/simulator.(*Service).call()
      /home/toomasp/tmp/govmomi/simulator/simulator.go:208 +0x5c0
  github.com/vmware/govmomi/simulator.(*Service).ServeSDK()
      /home/toomasp/tmp/govmomi/simulator/simulator.go:439 +0x16b5
  github.com/vmware/govmomi/simulator.(*Service).ServeSDK-fm()
      /home/toomasp/tmp/govmomi/simulator/simulator.go:399 +0x5f
  net/http.HandlerFunc.ServeHTTP()
      /opt/go/src/net/http/server.go:1995 +0x51
  net/http.(*ServeMux).ServeHTTP()
      /opt/go/src/net/http/server.go:2375 +0x28a
  net/http.serverHandler.ServeHTTP()
      /opt/go/src/net/http/server.go:2774 +0xce
  net/http.(*conn).serve()
      /opt/go/src/net/http/server.go:1878 +0x811

Previous write at 0x00c000776020 by goroutine 78:
  runtime.slicecopy()
      /opt/go/src/runtime/slice.go:197 +0x0
  github.com/vmware/govmomi/simulator.RemoveReference()
      /home/toomasp/tmp/govmomi/simulator/registry.go:396 +0x1ef
  github.com/vmware/govmomi/simulator.destroyView()
      /home/toomasp/tmp/govmomi/simulator/view_manager.go:74 +0x8a
  github.com/vmware/govmomi/simulator.(*ContainerView).DestroyView()
      /home/toomasp/tmp/govmomi/simulator/view_manager.go:144 +0x105
  runtime.call64()
      /opt/go/src/runtime/asm_amd64.s:520 +0x3a
  reflect.Value.Call()
      /opt/go/src/reflect/value.go:308 +0xc0
  github.com/vmware/govmomi/simulator.(*Service).call.func1()
      /home/toomasp/tmp/govmomi/simulator/simulator.go:209 +0x7b
  github.com/vmware/govmomi/simulator.(*Registry).WithLock()
      /home/toomasp/tmp/govmomi/simulator/registry.go:532 +0x51
  github.com/vmware/govmomi/simulator.(*Service).call()
      /home/toomasp/tmp/govmomi/simulator/simulator.go:208 +0x5c0
  github.com/vmware/govmomi/simulator.(*Service).ServeSDK()
      /home/toomasp/tmp/govmomi/simulator/simulator.go:439 +0x16b5
  github.com/vmware/govmomi/simulator.(*Service).ServeSDK-fm()
      /home/toomasp/tmp/govmomi/simulator/simulator.go:399 +0x5f
  net/http.HandlerFunc.ServeHTTP()
      /opt/go/src/net/http/server.go:1995 +0x51
  net/http.(*ServeMux).ServeHTTP()
      /opt/go/src/net/http/server.go:2375 +0x28a
  net/http.serverHandler.ServeHTTP()
      /opt/go/src/net/http/server.go:2774 +0xce
  net/http.(*conn).serve()
      /opt/go/src/net/http/server.go:1878 +0x811

Goroutine 54 (running) created at:
  net/http.(*Server).Serve()
      /opt/go/src/net/http/server.go:2884 +0x4c4
  github.com/vmware/govmomi/simulator/internal.(*Server).goServe.func1()
      /home/toomasp/tmp/govmomi/simulator/internal/server.go:268 +0xac

Goroutine 78 (running) created at:
  net/http.(*Server).Serve()
      /opt/go/src/net/http/server.go:2884 +0x4c4
  github.com/vmware/govmomi/simulator/internal.(*Server).goServe.func1()
      /home/toomasp/tmp/govmomi/simulator/internal/server.go:268 +0xac
==================
--- FAIL: TestViewManager_CreateContainerView (1.19s)
    testing.go:809: race detected during execution of test
FAIL
FAIL	github.com/vmware/govmomi/simulator	1.237s
```

--- a/simulator/session_manager.go
+++ b/simulator/session_manager.go
@@ -334,13 +334,20 @@ type Session struct {
 	*Registry
 }
 
-// Put wraps Registry.Put, setting the moref value to include the session key.
-func (s *Session) Put(item mo.Reference) mo.Reference {
+func (s *Session) setReference(item mo.Reference) {
 	ref := item.Reference()
 	if ref.Value == "" {
 		ref.Value = fmt.Sprintf("session[%s]%s", s.Key, uuid.New())
 	}
+	if ref.Type == "" {
+		ref.Type = typeName(item)
+	}
 	s.Registry.setReference(item, ref)
+}
+
+// Put wraps Registry.Put, setting the moref value to include the session key.
+func (s *Session) Put(item mo.Reference) mo.Reference {
+	s.setReference(item)
 	return s.Registry.Put(item)
 }
 

--- a/simulator/view_manager.go
+++ b/simulator/view_manager.go
@@ -69,10 +69,6 @@ func NewViewManager(ref types.ManagedObjectReference) object.Reference {
 }
 
 func destroyView(ref types.ManagedObjectReference) soap.HasFault {
-	m := Map.ViewManager()
-
-	RemoveReference(&m.ViewList, ref)
-
 	return &methods.DestroyViewBody{
 		Res: &types.DestroyViewResponse{},
 	}
@@ -118,8 +114,6 @@ func (m *ViewManager) CreateContainerView(ctx *Context, req *types.CreateContain
 	}
 
 	ctx.Session.setReference(container)
-
-	m.ViewList = append(m.ViewList, container.Reference())
 
 	body.Res = &types.CreateContainerViewResponse{
 		Returnval: container.Self,

--- a/simulator/view_manager.go
+++ b/simulator/view_manager.go
@@ -117,7 +117,7 @@ func (m *ViewManager) CreateContainerView(ctx *Context, req *types.CreateContain
 		}
 	}
 
-	ctx.Session.Put(container)
+	ctx.Session.setReference(container)
 
 	m.ViewList = append(m.ViewList, container.Reference())
 
@@ -127,6 +127,8 @@ func (m *ViewManager) CreateContainerView(ctx *Context, req *types.CreateContain
 
 	seen := make(map[types.ManagedObjectReference]bool)
 	container.add(root, seen)
+
+	ctx.Session.Registry.Put(container)
 
 	return body
 }

--- a/simulator/view_manager_test.go
+++ b/simulator/view_manager_test.go
@@ -111,16 +111,10 @@ func TestContainerViewVPX(t *testing.T) {
 
 	pc := property.DefaultCollector(c.Client)
 
-	mvm := Map.ViewManager()
-
 	for i, test := range tests {
 		cv, err := v.CreateContainerView(ctx, test.root, test.kinds, test.recurse)
 		if err != nil {
 			t.Fatal(err)
-		}
-
-		if len(mvm.ViewList) != 1 {
-			t.Errorf("ViewList=%s", mvm.ViewList)
 		}
 
 		var mcv mo.ContainerView
@@ -138,10 +132,6 @@ func TestContainerViewVPX(t *testing.T) {
 		err = cv.Destroy(ctx)
 		if err != nil {
 			t.Fatal(err)
-		}
-
-		if len(mvm.ViewList) != 0 {
-			t.Errorf("ViewList=%s", mvm.ViewList)
 		}
 	}
 }

--- a/simulator/view_manager_test.go
+++ b/simulator/view_manager_test.go
@@ -18,7 +18,10 @@ package simulator
 
 import (
 	"context"
+	"sync"
 	"testing"
+
+	"github.com/vmware/govmomi/vim25"
 
 	"github.com/vmware/govmomi"
 	"github.com/vmware/govmomi/object"
@@ -140,5 +143,68 @@ func TestContainerViewVPX(t *testing.T) {
 		if len(mvm.ViewList) != 0 {
 			t.Errorf("ViewList=%s", mvm.ViewList)
 		}
+	}
+}
+
+func TestViewManager_CreateContainerView(t *testing.T) {
+	m := VPX()
+	m.Datacenter = 10 // smaller numbers than this sometimes fail to trigger the DATA RACE
+	m.Datastore = 10
+	err := m.Run(func(ctx context.Context, client *vim25.Client) (err error) {
+		manager := view.NewManager(client)
+		datacenterView, err := manager.CreateContainerView(ctx, client.ServiceContent.RootFolder, []string{"Datacenter"}, true)
+		if err != nil {
+			return err
+		}
+		defer datacenterView.Destroy(ctx)
+		var datacenterList []mo.Datacenter
+		err = datacenterView.Retrieve(ctx, []string{"Datacenter"}, []string{"name", "datastoreFolder"}, &datacenterList)
+		if err != nil {
+			return err
+		}
+		var getDataStores = func(dc mo.Datacenter) (dsNames []string, err error) {
+			dataStoreView, err := manager.CreateContainerView(ctx, dc.DatastoreFolder, []string{"Datastore"}, true)
+			if err != nil {
+				return dsNames, err
+			}
+			defer dataStoreView.Destroy(ctx)
+			var dsList []mo.Datastore
+			err = dataStoreView.Retrieve(ctx, []string{"Datastore"}, []string{"name"}, &dsList)
+			if err != nil {
+				return dsNames, err
+			}
+			for _, ds := range dsList {
+				dsNames = append(dsNames, ds.Name)
+			}
+			return dsNames, err
+		}
+		wg := &sync.WaitGroup{}
+		mtx := sync.Mutex{}
+		var datastores [][]string
+		wg.Add(len(datacenterList))
+		for _, dc := range datacenterList {
+			go func(ref mo.Datacenter) {
+				defer wg.Done()
+				ds, err := getDataStores(ref)
+				if err != nil {
+					return
+				}
+				mtx.Lock()
+				datastores = append(datastores, ds)
+				mtx.Unlock()
+			}(dc)
+		}
+		wg.Wait()
+		if len(datastores) != m.Datacenter {
+			t.Errorf("Invalid number of datacenters: %d", len(datastores))
+		} else {
+			if len(datastores[0]) != m.Datastore {
+				t.Errorf("Invalid number of datastores per datacenter: %d", len(datastores[0]))
+			}
+		}
+		return nil
+	})
+	if err != nil {
+		t.Errorf("Failed to run simulation: %s", err.Error())
 	}
 }

